src/NetSparkle/AppCastHandlers/XMLAppCast.cs

```diff
@@ -182,66 +187,110 @@ protected virtual void ParseAppCast(string appCast)
+                return MatchingResult.NotThisPlatform;
+            }
 #endif
-                // filter smaller versions
-                if (new Version(item.Version).CompareTo(installed) <= 0)
+            // filter smaller versions
+            if (new Version(item.Version).CompareTo(installed) <= 0)
+            {
+                _logWriter.PrintMessage("Rejecting update for {0} ({1}, {2}) because it is older than our current version of {3}", item.Version,
+                    item.ShortVersion, item.Title, installed);
+                return MatchingResult.VersionIsOlderThanCurrent;
+            }
+            // filter versions without signature if we need signatures. But accept version without downloads.
+            if (signatureNeeded && string.IsNullOrEmpty(item.DownloadSignature) && !string.IsNullOrEmpty(item.DownloadLink))
+            {
+                _logWriter.PrintMessage("Rejecting update for {0} ({1}, {2}) because it we needed a DSA/other signature and " +
+                    "the item has no signature yet has a download link of {3}", item.Version,
+                    item.ShortVersion, item.Title, item.DownloadLink);
+                return MatchingResult.SignatureIsMissing;
+            }
+
+            return MatchingResult.Valid;
+        }
+
+        /// <summary>
+        /// Returns sorted list of updates between current installed version and latest version in <see cref="Items"/>. 
+        /// Currently installed version is NOT included in the output.
+        /// </summary>
+        /// <returns>A list of <seealso cref="AppCastItem"/> updates that could be installed</returns>
+        public virtual List<AppCastItem> GetAvailableUpdates()
+        {
+            Version installed = new Version(_config.InstalledVersion);
+            List<AppCastItem> appCastItems = Items;
+
+            if (AppCastFilter != null)
+            {
+                var result = AppCastFilter.GetFilteredAppCastItems(installed, Items);
+                if (result.FilteredAppCastItems != null)
                 {
-                    _logWriter.PrintMessage("Rejecting update for {0} ({1}, {2}) because it is older than our current version of {3}", item.Version,
-                        item.ShortVersion, item.Title, installed);
-                    return false;
+                    if (result.ForceInstallOfLatestInFilteredList)
+                    {
+                        // if there is just a single version in the FilterResult; we'll fake up a version number - so that the
+                        // NetSparkle code will request a re-install of that version (and it'll shown only a single .md file in the changelog)
+                        if (result.FilteredAppCastItems.Count == 1)
+                            installed = new Version("0.0.0");
+                        else if (result.FilteredAppCastItems.Count > 1)
+                            installed = new Version(result.FilteredAppCastItems[1].Version);
```

Comment:
Can you help me understand what this line is doing, please? Why are you grabbing the second item in the list (index 1) if there is more than 1 item? For grabbing a version earlier than the first one? What happens if this item is the same version as `FilteredAppCastItems[0]` but for a different OS?

Reply:
I picked 0.0.0 because I really wanted to do the following: result.FilteredAppCastItems[0] - 'a version', e.g. I wanted to go back 1 version step to ensure that the item that was found was identified as the version that must be reinstalled.  I don't believe there is a minus operator for the semantic version class - that would have done the trick (a bit like date arithmetic, pick a date take 1 day off it).

The reasoning is that in this case the user is moving from beta back to stable - so no matter what, you want to force the installation of the latest version of the stable app.  That means "just get me the latest non beta" and install it.  In the case where there is just 1 appcast item; that's the one we want to install.  

Without the 0.0.0; the code presently discards this appcastitem because its the same as the currently installed one and therefore I forced the installed version to 0.0.0 which in turn will cause the latest discovered item to be the target for an install.

If you can think of a better way to achieve that - i'm all for it.  the 0.0.0 gets it done, but your instinct is good - its a bit of a code smell and takes a few sentences here to describe why (bad sign - nobody else will find/read these comments).

Reply:
I have tried a different approach - the unit tests still pass but I have not tested it live with my own s/ware yet.  I would like to do that as well before merging.  

I now pass a flag into the IsMatchingUpdate() method to indicate whether or not the appcastitem should be checked against the installed version - this means its possible to delete the hacky code.

Let me know what you think could be improved.